workers/unity/Packages/io.improbable.gdk.core/Systems/EntitySystem.cs

```diff
@@ -1,61 +1,127 @@
-                foreach (var entityId in diff.GetEntitiesAdded())
+            public NativeSlice<EntityId> Slice
+            {
+                get
                 {
-                    entitiesAdded.Add(entityId);
+                    JobHandle.Complete();
+                    return new NativeSlice<EntityId>(EntityIds, 0, EntityCount);
                 }
+            }
 
-                foreach (var entityId in diff.GetEntitiesRemoved())
+            public void EnsureSize(ref EntityQuery query)
+            {
+                EntityCount = query.CalculateEntityCount();
+                if (EntityIds.Length >= EntityCount)
                 {
-                    entitiesRemoved.Add(entityId);
+                    return;
                 }
 
-                if (entitiesAdded.Count != 0 || entitiesRemoved.Count != 0)
+                EntityIds.Dispose();
+                EntityIds = new NativeArray<EntityId>(EntityCount << 1, Allocator.Persistent);
+            }
+
+            public void Dispose()
+            {
+                if (EntityIds.IsCreated)
                 {
-                    ViewVersion += 1;
+                    EntityIds.Dispose();
                 }
             }
         }
 
+        private EndSimulationEntityCommandBufferSystem bufferSystem;
+        private EntityCollection added;
+        private EntityQuery addedQuery; // This query needs to be a member of SystemBase
+        private EntityCollection removed;
+        private EntityQuery removedQuery; // This query needs to be a member of SystemBase
+        public int ViewVersion { get; private set; }
+        public NativeSlice<EntityId> EntitiesRemoved => removed.Slice;
+        public NativeSlice<EntityId> EntitiesAdded => added.Slice;
+
+        protected override void OnCreate()
+        {
+            base.OnCreate();
+            bufferSystem = World.GetOrCreateSystem<EndSimulationEntityCommandBufferSystem>();
+            removed = new EntityCollection(1);
+            added = new EntityCollection(1);
+        }
+
         protected override void OnUpdate()
         {
-            entitiesAdded.Clear();
-            entitiesRemoved.Clear();
+            UpdateEntitiesAdded();
+            UpdateEntitiesRemoved();
+
+            Dependency = JobHandle.CombineDependencies(added.JobHandle, removed.JobHandle);
+            bufferSystem.AddJobHandleForProducer(Dependency);
+
+            if (added.EntityCount != 0 || removed.EntityCount != 0)
+            {
+                ViewVersion += 1;
+            }
+        }
+
+        private void UpdateEntitiesAdded()
+        {
+            added.EnsureSize(ref addedQuery);
+            var buffer = bufferSystem.CreateCommandBuffer().AsParallelWriter();
+            var array = added.EntityIds;
+            added.JobHandle = Entities.WithName("EntitiesAdded")
+                .WithNone<EntitySystemStateComponent>()
+                .WithAll<SpatialEntityId>()
```

Comment:
No need for this, as the constraint is already created with the `in SpatialEntityId entityId`